Implement network plugin capabilities hook and shaping capability

dcbw opened the pull request:
Allow network plugins to declare that they handle shaping and that
Kuberenetes should not.  I've got an OpenShift PR that handles shaping in OVS but the kubelet code sends nasty pod events because it doesn't think shaping is used since --reconcile-cbr0 is not used with most network plugins.

See https://github.com/openshift/openshift-sdn/pull/266 for the OpenShift implementation.

dcbw commented:
@thockin one more for your pile...

vishh commented:
LGTM

wojtek-t commented:
@dcbw - this PR doesn't compile:

# k8s.io/kubernetes/pkg/kubelet/network/kubenet
_output/dockerized/go/src/k8s.io/kubernetes/pkg/kubelet/network/kubenet/kubenet_unsupported.go:32: cannot use kubenetNetworkPlugin literal (type *kubenetNetworkPlugin) as type network.NetworkPlugin in return argument:
	*kubenetNetworkPlugin does not implement network.NetworkPlugin (missing Capabilities method)

Can you please fix?

dcbw commented:
@wojtek-t right you are, sorry about that, should be fixed now.

dcbw commented:
Jenkins test flake is https://github.com/kubernetes/kubernetes/issues/22856

vishh commented on pkg/kubelet/network/plugins.go:
```diff
@@ -54,6 +60,9 @@ type NetworkPlugin interface {
 	Name() string
 
+	// Returns a bitfield of plugin capabilities
+	Capabilities() uint
```
nit: I prefer returning a `Set` instead of `uint`.

dcbw replied:
> nit: I prefer returning a Set instead of uint.

Do you mean an Int Set or a String Set?  I only get one hit for Int sets (pkg/client/restclient/urlbackoff.go) so it seems like there aren't a ton of Int/Int64/Byte Set users yet.  I'm OK moving to a Set though if you want one.

vishh replied:
A String Set is what I meant. An Int Set might be better though for this
use case.

On Fri, Mar 11, 2016 at 9:55 AM, Dan Williams <notifications@github.com>
wrote:

> In pkg/kubelet/network/plugins.go
> <https://github.com/kubernetes/kubernetes/pull/22827#discussion_r55863526>
> :
>
> > @@ -54,6 +60,9 @@ type NetworkPlugin interface {
> >  	// for a plugin by name, e.g.
> >  	Name() string
> >
> > +	// Returns a bitfield of plugin capabilities
> > +	Capabilities() uint
>
> nit: I prefer returning a Set instead of uint.
>
> Do you mean an Int Set or a String Set? I only get one hit for Int sets
> (pkg/client/restclient/urlbackoff.go) so it seems like there aren't a ton
> of Int/Int64/Byte Set users yet. I'm OK moving to a Set though if you want
> one.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/22827/files#r55863526>.
>

dcbw commented:
@wojtek-t @vishh PTAL, thanks!

bprashanth commented on pkg/kubelet/kubelet.go:
```diff
@@ -1795,19 +1795,21 @@ func (kl *Kubelet) syncPod(pod *api.Pod, mirrorPod *api.Pod, podStatus *kubecont
-		} else if kl.shaper != nil {
-			if len(apiPodStatus.PodIP) > 0 {
-				err = kl.shaper.ReconcileCIDR(fmt.Sprintf("%s/32", apiPodStatus.PodIP), egress, ingress)
+	if kl.shapingEnabled() {
```
Can you avoid the extra indent and just return nil if !kl.shapingEnabled()?

bprashanth commented on pkg/kubelet/network/kubenet/kubenet_linux.go:
```diff
@@ -177,6 +177,10 @@ func (plugin *kubenetNetworkPlugin) Name() string {
 	return KubenetPluginName
 }
 
+func (plugin *kubenetNetworkPlugin) Capabilities() uint {
```
instead of implementing this method over and over can you just embed a struct like:
```go
type pluginCapabilities uint

func (pc *pluginCapabilities) BWShaping() bool {
  return pc & NET_PLUGIN_CAPABILITY_SHAPING
}

 type kubenetNetworkPlugin struct {
  pluginCapabilities
 }
 
 func NewPlugin() network.NetworkPlugin {
  return &kubenetNetworkPlugin{pluginCapabilities: 0}
 }
```

dcbw replied:
@bprashanth I think this approach is more "magic", unless I don't quite understand what you're suggesting.  If the goal is to get rid of Capabilities(), I think it's better to have a well-defined API between the plugins and the core and keep an explicit method.  Getting rid of Capabilities() and embedding the capabilities in the plugin type would mean every plugin has to embed the capabilities and we lose some of the benefit of plugins being an interface.  Does that make sense?

dcbw replied:
We could however export NoopNetworkPlugin and then embed that in all the other plugins, and that would get rid of the Capabilities() and Event() methods in them.

bprashanth commented on pkg/kubelet/network/plugins.go:
```diff
@@ -166,6 +175,10 @@ func (plugin *noopNetworkPlugin) Name() string {
 }
 
+func (plugin *noopNetworkPlugin) Capabilities() uint {
+	return 0
```
Souldn't this actually say that shaping is enabled to support existing behavior?

dcbw replied:
The noop plugin itself doesn't support shaping, but the kubelet core does completely outside of the plugins.  The capability is there to indicate that the plugin itself handles shaping and the core doesn't need to.

bprashanth commented:
@kubernetes/goog-cluster

dcbw committed:
Embed NoopNetworkPlugin into other network plugins

Let Noop handle common functions.

dcbw committed:
Implement network plugin capabilities hook and shaping capability

Allow network plugins to declare that they handle shaping and that
Kuberenetes should not.  Will be first used by openshift-sdn which
handles shaping through OVS, but this triggers a warning when
kubelet notices the bandwidth annotations.

dcbw commented:
@bprashanth @vishh PTAL, thanks!

vishh commented:
@k8s-bot test node e2e experimental

vishh commented on pkg/util/sets/uint.go:
```diff
@@ -0,0 +1,194 @@
+/*
```
nit: Kindly adding relevant unit tests for this library.

dcbw replied:
> nit: Kindly adding relevant unit tests for this library.

@vishh added and repushed.  PTAL thanks!

bprashanth commented on pkg/kubelet/kubelet.go:
```diff
@@ -3591,6 +3597,15 @@ func (kl *Kubelet) updatePodCIDR(cidr string) {
+
+func (kl *Kubelet) shapingEnabled() bool {
+	// Disable shaping if a network plugin is defined and supports shaping
+	if kl.networkPlugin != nil && kl.networkPlugin.Capabilities().Has(network.NET_PLUGIN_CAPABILITY_SHAPING) {
```
nit: can you just return this?

dcbw replied:
Not sure what you mean here...

bprashanth replied:
return kl.networkPlugin != nil && kl.networkPlugin.Capabilities().Has(network.NET_PLUGIN_CAPABILITY_SHAPING)

Also why return false if the network plugin has shaping enabled?

bprashanth replied:
oh, it just read weird, this is answering the question is shaping enabled inline for the kubelet. Please add comment saying that the plugin will handle in, for hasty people like myself.

bprashanth commented on pkg/util/sets/uint.go:
```diff
@@ -0,0 +1,194 @@
+Copyright 2016 The Kubernetes Authors All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
```
do we really need this, we already have an intset right? why not just use plain ints for declaring capabilities?

dcbw replied:
> do we really need this, we already have an intset right? why not just use plain ints for declaring capabilities?

Seems pretty odd to allow negative numbers for something that's really just a set of flags, that's all.

bprashanth replied:
Less code is better ? 
Also x := 10 will give you an int, so if we ever want to do an exact compare anywhere uint would lead to unnecessary casting

dcbw replied:
Ok, shall use int sets then.

dcbw commented:
@bprashanth PTAL, thanks!

bprashanth commented:
LGTM, thanks

freehan commented on pkg/kubelet/network/kubenet/kubenet_linux.go:
```diff
@@ -41,6 +41,8 @@ const (
 )
 
 type kubenetNetworkPlugin struct {
+	network.NoopNetworkPlugin
```
It looks to me that kubenet already support shaping
https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/network/kubenet/kubenet_linux.go#L211

Just include NET_PLUGIN_CAPABILITY_SHAPING for kubenet would work

freehan replied:
Please ignore my comment above. The shaper in kubenet is not complete. Ingress/egress for pod has not been covered.
